### PR TITLE
More CAP-38 path payment tests

### PR DESCRIPTION
**Pull request by jonjove**
# Description
Built on top of #3198, so that should be merged first. New code starts from "CAP-38: Fix bug with PATH_PAYMENT_STRICT_*_CROSS_SELF".

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/transactions/PathPaymentOpFrameBase.cpp [hunk 1]
```diff
@@ -81,13 +81,17 @@ PathPaymentOpFrameBase::convert(AbstractLedgerTxn& ltx,
...
             {
                 // we are crossing our own offer
-                setResultOfferCrossSelf();
                 return OfferFilterResult::eStop;
             }
             return OfferFilterResult::eKeep;
         },
         offerTrail, maxOffersToCross);
 
+    if (r == ConvertResult::eFilterStop)
```
Comment (MonsieurNicolas) [inline src/transactions/PathPaymentOpFrameBase.cpp:90]: should this just be moved to the switch below (line 103 in this commit)?

**Review by MonsieurNicolas — CHANGES_REQUESTED** [review]

File: src/transactions/PathPaymentOpFrameBase.cpp (hunk 1 above)
Comment (jonjove) [reply src/transactions/PathPaymentOpFrameBase.cpp:90]: I considered that but it's _technically_ a protocol change to do that. I'm comfortable with it if you are, so let me know.

Comment (MonsieurNicolas) [reply src/transactions/PathPaymentOpFrameBase.cpp]: I don't understand why this would be a protocol change: the only difference in this function is that in case of exception below the result would not be set - but in that case, we end up with an internal error anyways.

File: src/transactions/ManageOfferOpFrameBase.cpp [hunk 2]
```diff
@@ -367,7 +372,8 @@ ManageOfferOpFrameBase::doApply(AbstractLedgerTxn& ltxOuter)
         case ConvertResult::ePartial:
             sheepStays = true;
             break;
-        case ConvertResult::eFilterStop:
+        case ConvertResult::eFilterStopBadPrice:
+        case ConvertResult::eFilterStopCrossSelf:
```
Comment (MonsieurNicolas) [inline src/transactions/ManageOfferOpFrameBase.cpp:376]: You should handle those two cases separately and remove the `if (!isResultSuccess())` (this was used to distinguish between `eFilterStopBadPrice` and `eFilterStopCrossSelf`), that way you can also move `setResultCrossSelf` into the `eFilterStopCrossSelf` case

File: src/transactions/PathPaymentOpFrameBase.cpp [hunk 3]
```diff
@@ -99,7 +99,8 @@ PathPaymentOpFrameBase::convert(AbstractLedgerTxn& ltx,
 
     switch (r)
     {
-    case ConvertResult::eFilterStop:
+    case ConvertResult::eFilterStopBadPrice:
```
Comment (MonsieurNicolas) [inline src/transactions/PathPaymentOpFrameBase.cpp:102]: why handle `eFilterStopBadPrice` like this? I think it should just hit the `default` and crash as this should never happen

**Review by MonsieurNicolas — CHANGES_REQUESTED** [review]
couple questions there

File: src/transactions/PathPaymentOpFrameBase.cpp (hunk 1 above)
Comment (jonjove) [reply src/transactions/PathPaymentOpFrameBase.cpp]: Oh yes, you're right. I'll make this change.

**Comment by MonsieurNicolas** [comment]
r+ ecfdeb1081742ee2b6b7bd8463b213f22c3be691

File: src/transactions/test/LiquidityPoolTradeTests.cpp [hunk 4]
```diff
@@ -21,6 +21,957 @@
...
+                                 int64_t amountB, Price const& minPrice,
+                                 Price const& maxPrice) {
+        if (!(cur1 < cur2))
+        {
+            std::swap(amountA, amountB);
+        }
+        acc.liquidityPoolDeposit(pool12, amountA, amountB, minPrice, maxPrice);
+    };
+
+    SECTION("chooses best price")
```
[the remark this answers is not in the document]
Comment (jonjove) [reply src/transactions/test/LiquidityPoolTradeTests.cpp:58]: Done. It wasn't super easy because there are several things to handle, but I really like that it applies to all our tests globally.

[the remark this answers is not in the document]
Comment (jonjove) [reply src/transactions/test/LiquidityPoolTradeTests.cpp:58]: (To be clear, done in my new work. Not done in this PR.)

[the remark this answers is not in the document]
Comment (jonjove) [reply src/transactions/test/LiquidityPoolTradeTests.cpp:58]: There were some edge cases I couldn't sort out easily. It almost works, but I'll have to revisit it later.
